internal/pkg/manifest/env.go

```diff
@@ -102,9 +102,56 @@ type environmentNetworkConfig struct {
 }
 
 type environmentVPCConfig struct {
-	ID      *string              `yaml:"id,omitempty"`
-	CIDR    *IPNet               `yaml:"cidr,omitempty"`
-	Subnets subnetsConfiguration `yaml:"subnets,omitempty"`
+	ID                  *string              `yaml:"id,omitempty"`
+	CIDR                *IPNet               `yaml:"cidr,omitempty"`
+	Subnets             subnetsConfiguration `yaml:"subnets,omitempty"`
+	SecurityGroupConfig securityGroupConfig  `yaml:"security_group,omitempty"`
+}
+
+type securityGroupConfig struct {
+	Ingress []securityGroupRule `yaml:"ingress,omitempty"`
+	Egress  []securityGroupRule `yaml:"egress,omitempty"`
+}
+
+// IsEmpty returns true if there is no security group rule defined.
+func (cfg securityGroupConfig) IsEmpty() bool {
+	return len(cfg.Ingress) == 0 && len(cfg.Egress) == 0
+}
+
+type securityGroupRule struct {
+	CidrIP     string `yaml:"cidr"`
+	FromPort   int    `yaml:"from_port"`
+	IpProtocol string `yaml:"ip_protocol"`
+	ToPort     int    `yaml:"to_port"`
+}
+
+//IsValid checks for valid security group config.
+func (cfg securityGroupConfig) IsValid() bool {
+	for _, ingress := range cfg.Ingress {
+		if ingress.IsEmpty() {
+			return false
+		}
+	}
+	for _, egress := range cfg.Egress {
+		if egress.IsEmpty() {
+			return false
+		}
+	}
+	return true
+}
+
+// EnvSecurityGroup returns the security group config if the user has set any values.
+// If there is no env security group settings, then returns nil and false.
+func (cfg *EnvironmentConfig) EnvSecurityGroup() (*securityGroupConfig, bool) {
+	if isEmpty := cfg.Network.VPC.SecurityGroupConfig.IsEmpty(); !isEmpty {
+		return &cfg.Network.VPC.SecurityGroupConfig, true
+	}
+	return nil, false
+}
+
+//IsEmpty checks for required parameters in the security group rules.
+func (cfg securityGroupRule) IsEmpty() bool {
+	return cfg.CidrIP == "" || cfg.IpProtocol == ""
```

Comment:
Should this be an `&&`?

Reply:
None of these should be empty as they are required parameters hence `or` is required

Reply:
"They shouldn't be empty" this should have been handled in `Validate`. `Validate` should validate that any required field is not empty.

What you have here would result in the following:
```
c := securityGroupRule {
   CidrIP: "ip"
}
// c.IsEmpty() is true
```
which seems wrong because `securityGroupRule` is apparently non-empty 😢  

I feel like we should make a distinction between "what is required" and "what is empty". Missing a required field doesn't mean the structure is empty; instead, i think it means the input is invalid.

Reply:
As per the discussion renaming this method as `Validate() error` for the less confusion.